pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java

```diff
@@ -16,130 +16,85 @@
 
 package com.linkedin.pinot.common.query.context;
 
+import com.google.common.base.Preconditions;
 import com.linkedin.pinot.common.metrics.ServerMetrics;
 import com.linkedin.pinot.common.metrics.ServerQueryPhase;
 import com.linkedin.pinot.common.request.BrokerRequest;
-import java.util.HashMap;
-import java.util.Map;
 import java.util.concurrent.TimeUnit;
-import javax.annotation.Nullable;
 
-public class TimerContext {
 
-  private final ServerMetrics serverMetrics;
-  private final Map<ServerQueryPhase, Timer> phaseTimers = new HashMap<>();
-  private final BrokerRequest brokerRequest;
-  private long queryArrivalTimeNs;
+public class TimerContext {
+  private final BrokerRequest _brokerRequest;
+  private final ServerMetrics _serverMetrics;
+  private final long _queryArrivalTimeNs;
+  private final Timer[] _phaseTimers = new Timer[ServerQueryPhase.values().length];
 
-  public class Timer implements AutoCloseable {
-    private final ServerQueryPhase queryPhase;
-    long startTimeNs;
-    long endTimeNs;
-    boolean isRunning;
+  public class Timer {
+    private final ServerQueryPhase _queryPhase;
+    private final long _startTimeNs;
 
-    Timer(ServerQueryPhase phase) {
-      this.queryPhase = phase;
-    }
+    private long _durationNs;
+    private boolean _ended;
 
-    void start() {
-      startTimeNs = System.nanoTime();
-      isRunning = true;
-    }
-
-    void setStartTimeNs(long startTimeNs) {
-      this.startTimeNs = startTimeNs;
-      isRunning = true;
+    private Timer(ServerQueryPhase queryPhase, long startTimeNs) {
+      _queryPhase = queryPhase;
+      _startTimeNs = startTimeNs;
     }
 
     public void stopAndRecord() {
-      if (isRunning) {
-        endTimeNs = System.nanoTime();
-
-        recordPhaseTime(this);
-        isRunning = false;
-      }
-    }
-
-    public long getDurationNs() {
-      return endTimeNs - startTimeNs;
+      Preconditions.checkState(!_ended, "Timer has already stopped for query phase: " + _queryPhase);
```

Comment:
I think it is OK to not do a Preconditions check here. We should not fail a query because of some bug in our code that calls stop() twice. I would suggest to replace this with a warning and return

Reply:
Fixed.